### ISSUE TITLE
Lab2 review

### ISSUE THREAD

**Issue by GiorgioBongiovanni** [vitosilver99/CI2024_lab2#1]
First of all, I want to congratulate you on the excellent results you have achieved.
I haven’t read a description from you but I believe you implemented the order crossover as the type of crossover operator and I think this was a key choice in achieving such high accuracy compared to the optimal results for the following reasons.
The TSP imposes specific constraints that any solution must adhere to:
- Valid Permutation: Each city must be visited exactly once.
- Adjacency: The order in which cities are visited determines the solution's quality since it affects the total distance.
- Cyclic Structure: The first and last cities must be connected to close the cycle.

The Order Crossover is designed to respect these constraints making it a natural choice for the TSP.
Order Crossover operates in two stages:
1) Preservation of Partial Order: A segment from the first parent is copied into the child at the same position.
2) Filling Missing Positions: The child is completed with the remaining cities from the second parent maintaining their relative order.
This strategy ensures that the cities copied from the first parent maintain their order and local adjacencies and the cities from the second parent contribute diversity by introducing new connections.
Order Crossover is particularly well-suited for the TSP for several reasons, explained in relation to the problem's constraints and goals:
Firstly, the segment copied from the first parent remains intact in the child preserving local adjacencies.
This is crucial in the TSP because good solutions often emerge from configurations that retain favorable connections between neighboring cities.
Secondly, by filling the remaining positions with cities from the second parent, Order Crossover introduces new connections and increases diversity.
This balance between exploitation (preserving good structures) and exploration (introducing new configurations) helps avoid local optima.
Thirdly, Order Crossover always produces valid permutations, where each city appears exactly once and, while the TSP requires a cyclic structure, Order Crossover inherently manages the connection between the first and last cities, as the child is constructed to respect the parents' relative order.

A suggestion I would like to propose to improve the computation time of the genetic algorithm is the following:
Introducing an early stopping criterion based on the stagnation of the best individual's fitness is a practical strategy to improve the efficiency of the genetic algorithm. 
If the fitness improvement stalls, the algorithm is likely close to a local minimum or the optimal solution thus continuing to generate new populations may not yield significant benefits.
For this reason, lack of fitness improvement over several generations is a sign that the population has reached a stable state.
This approach could avoid wasting computational resources on unnecessary iterations, especially for complex problems like the TSP.
The idea is to monitor the fitness improvement of the best individual. If the fitness does not improve for a predetermined number of consecutive generations (for example, 50 generations), the algorithm terminates:
``` python
no_improvement_count = 0
max_no_improvement = 50  # Maximum generations without fitness improvement

best_fitness = float('-inf')  # Initialize best fitness to the lowest possible value

for generation in range(MAX_GENERATIONS):
    # Execute the steps of the genetic algorithm
    # (e.g., selection, crossover, mutation, and fitness evaluation)
    ...

    # Find the fitness of the best individual in the current generation
    current_best_fitness = max(individual.fitness for individual in population)

    # Check if there is an improvement in the best fitness
    if current_best_fitness > best_fitness:  # Fitness must increase to improve
        best_fitness = current_best_fitness  # Update the best fitness
        no_improvement_count = 0  # Reset the no improvement counter
        print(f"Generation {generation}: Fitness improved to {best_fitness:.2f}")
    else:
        no_improvement_count += 1  # Increment the no improvement counter
        print(f"Generation {generation}: No improvement for {no_improvement_count} generations.")

    # Terminate if no improvement has been observed for the specified limit
    if no_improvement_count >= max_no_improvement:
        print(f"Terminating early at generation {generation} due to stagnation.")
        break

# Final output of the best solution
print(f"Algorithm terminated. Best fitness found: {best_fitness:.2f}")

```
From my point of view, instead of only tracking the best individual, you could also monitor the average fitness improvement of the entire population. This provides a more comprehensive picture of the algorithm's convergence.

In conclusion, it is likely that Order Crossover played a key role in preserving and combining the essential properties of the parents, leading to high-quality solutions. Good job!
